### PR TITLE
feat: add audio context + styling

### PR DESCRIPTION
**Pull request by akshaynthakur**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :cloud: :cloud: Description :cloud: :cloud:
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
This PR implements and Audio Context & Provider, most of the desired Play Screen functionality with hard-coded audio data, as well as all of the styling for the play screen.
- Audio play/pause toggle button
- ~~Audio skip backward button (currently just restarts audio)~~
- Audio fast-forward 30 seconds button
- Audio rewind 30 seconds button
- Audio thumbnail/image
- Save audio modal (localStorage functionality not implemented)
- Share audio modal with functionality: currently shares the phrase "Check out this podcast from Cultural Survival!" along with the SC url for the audio

### Screenshots :calling:
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @leexesther at the bottom."

Design updates: 
- Removed seek bar for now, replaced with 2 new buttons: one for fast-forwarding 30 seconds and one for rewinding 30 seconds
- Removed skip track buttons because not implementing audio queue
- Created my own pause button icon in Figma, please lmk if you want me to replace it with something you make
CC: @leexesther

https://user-images.githubusercontent.com/78326649/229383804-097a61ba-0ed8-4ae2-bbbf-05121cdfd502.mp4

## :zap: How to review :zap:
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

- Review all at once
- Check styling: is everything lined up? centered when it should be? spacing/margins feel right?
- Try different sequences of playing/pausing/forwarding etc. the audio to make sure all buttons work as intended

## :seedling: Next steps :seedling:
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
- ~~Audio skip forward button (not implemented currently)~~
- End of track behavior is currently undefined:
  - need listener of some sort maybe? detect end of track and update AudioContext approporiately
  - Audio fast forward: if +30 seconds goes past audio length, update AudioContext similarly
- Save audio modal (localStorage functionality not implemented)
- Implement dynamic seek bar (stretch)
- ~~Text horizontal scroll to keep title and artist textbox in one line in case of very long title/artist?~~ (implemented)
- Navigation: bottom navbar vs. down-arrow at top of screen
- Down-arrow header text: audio title or some other text?
- ~~audio plays only when the ringer is on for ios~~
- ~~playing the audio, leaving the screen, entering the screen, and playing again starts two tracks, and you lose control of the first audio~~ (fixed with AudioContext)

### Online sources :information_source:
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

### Related PRs :raised_hands:
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @davidqing6432

### CODE REVIEW

File: src/i18n/i18n.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ const createI18n = (language: string): i18nInstance => {
   const i18n = i18next.createInstance().use(initReactI18next);
 
   i18n.init({
+    compatibilityJSON: 'v3',
```
Comment (david-qing) [inline src/i18n/i18n.ts:10]: is this an intended change?

**Comment by akshaynthakur** [comment]
Not sure why eslint/prettier is failing on the 2 most recent commits; no issues/red squiggles show up for me in VSCode.

File: src/i18n/i18n.ts (hunk 1 above)
Comment (akshaynthakur) [reply src/i18n/i18n.ts:10]: yes! makes i18n android compatible.

https://stackoverflow.com/questions/70493788/i18nextpluralresolver-your-environment-seems-not-to-be-intl-api-compatible-u

Comment (david-qing) [reply src/i18n/i18n.ts:10]: got it!!!

File: src/screens/PlayScreen/Play.tsx [hunk 2]
```diff
@@ -1,69 +1,390 @@
...
+  Modal,
+  Pressable,
+  Text,
+  TouchableWithoutFeedback,
+  View,
+} from 'react-native';
+import { Audio } from 'expo-av';
+import React, { useState } from 'react';
+import Svg, { Circle, Path, Rect } from 'react-native-svg';
+// import Clipboard from '@react-native-clipboard/clipboard';
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: do we still need the clipboard comments? or is this for testing

File: src/screens/PlayScreen/Play.tsx [hunk 3]
```diff
@@ -1,69 +1,390 @@
...
 import Icon from '../../../assets/icons';
-import Colors from '../../styles/Colors';
 import styles from './styles';
 
+function play() {
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: can we extract these icons into the icons folder? or will we have to use different packages?

File: src/screens/PlayScreen/Play.tsx [hunk 4]
```diff
@@ -1,69 +1,390 @@
...
+    'https://i1.sndcdn.com/artworks-jeSDFXAMxLeFlfXx-a4zovA-t500x500.jpg',
+  artist: 'Jennifer Tauli',
+  title:
+    'Corpuz Talks about the Important Factors for Indigenous Peoples at COP15',
+  scLink:
+    'https://soundcloud.com/culturalsurvival/jennifer-tauli-corpuz-talks-about-the-important-factors-for-indigenous-peoples-at-cop15',
+};
+
 function PlayScreen() {
+  const [playState, setPlayState] = useState({
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: maybe we can create a play state type? is that better for htis

File: src/screens/PlayScreen/Play.tsx [hunk 5]
```diff
@@ -1,69 +1,390 @@
...
+    isSeeking: false,
+  });
+
+  async function toggleAudio(
+    sound: React.MutableRefObject<Audio.Sound>,
+    url: string,
+  ) {
+    const result = await sound.current.getStatusAsync();
+    if (result.isLoaded) {
+      if (result.isPlaying === false) {
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: might be a little more concise to write result.isPlaying (and then switch the two blocks)? or does this cause bugs

File: src/screens/PlayScreen/Play.tsx (hunk 2 above)
Comment (akshaynthakur) [reply src/screens/PlayScreen/Play.tsx]: clipboard/audio share functionality included in next steps

**Comment by david-qing** [comment]
Text horizontal scroll to keep title and artist textbox in one line in case of very long title/artist?
- this is a great idea! (we will probably also have a preview page like spotify podcasts, so if they need to see the full thing they'll look there)

Navigation: bottom navbar vs. down-arrow at top of screen
- we should implement a stack on the root so that we have the play screen pop up over the navigation bar

File: assets/icons.tsx [hunk 6]
```diff
@@ -112,37 +111,27 @@ const IconSvgs: Record<IconType, React.ReactElement> = {
...
   ),
   audio_forward: (
     <SvgXml
       xml={`
-    <svg width="44" height="29" viewBox="0 0 49 34" fill="none" xmlns="http://www.w3.org/2000/svg">
-<path d="M31 15.268C32.3333 16.0378 32.3333 17.9622 31 18.732L11.5 29.9904C10.1667 30.7602 8.5 29.7979 8.5 28.2583V5.74167C8.5 4.20207 10.1667 3.23982 11.5 4.00962L31 15.268Z" fill="#D9D9D9"/>
-<path d="M46 15.268C47.3333 16.0378 47.3333 17.9622 46 18.732L26.5 29.9904C25.1667 30.7602 23.5 29.7979 23.5 28.2583V5.74167C23.5 4.20207 25.1667 3.23982 26.5 4.00962L46 15.268Z" fill="#D9D9D9"/>
-</svg>
+      <svg width="36" height="34" viewBox="0 0 36 34" fill="none" xmlns="http://www.w3.org/2000/svg">
+      <path d="M31 15.268C32.3333 16.0378 32.3333 17.9622 31 18.732L11.5 29.9904C10.1667 30.7602 8.5 29.7979 8.5 28.2583V5.74167C8.5 4.20207 10.1667 3.23982 11.5 4.00962L31 15.268Z" fill="#CC502F"/>
```
Comment (david-qing) [inline assets/icons.tsx]: all fill colors in the icons file should be using the colors from global colors (in styles/colors.ts)

File: src/screens/PlayScreen/Play.tsx [hunk 7]
```diff
@@ -1,69 +1,390 @@
...
+  }
+
+  async function restartAudio(sound: React.MutableRefObject<Audio.Sound>) {
+    await sound.current.setPositionAsync(0);
+  }
+
+  async function rewindAudio(sound: React.MutableRefObject<Audio.Sound>) {
+    const result = await sound.current.getStatusAsync();
+    if (result.isLoaded) {
+      await sound.current.setPositionAsync(result.positionMillis - 30000);
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: maybe we can have the 30 seconds not hardcoded, instead pass in a milliseconds parameter into the function

**Comment by david-qing** [comment]
after testing, it seems like the audio plays only when the ringer is on for ios. this is fine for now, but add this to next steps

**Comment by david-qing** [comment]
also, playing the audio, leaving the screen, entering the screen, and playing again starts two tracks, and you lose control of the first audio. this should be added as a noted bug/next step

File: src/screens/PlayScreen/Play.tsx [hunk 8]
```diff
@@ -1,69 +1,390 @@
...
+        onRequestClose={() => {
+          setAudioModalVisible(!audioModalVisible);
+        }}
+      >
+        <View style={styles.modal}>
+          <View style={{ alignItems: 'center', marginTop: 305 }}>
+            <View style={styles.inset}>
+              <Text style={styles.inset_text1}>
+                {audioSaved ? 'Saved to Library!' : 'Removed from Library!'}
+              </Text>
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: we should add a horizontal line to separate the "saved to library" and "ok"

File: src/screens/PlayScreen/Play.tsx [hunk 9]
```diff
@@ -1,69 +1,390 @@
...
+      <Modal
+        animationType="fade"
+        transparent
+        visible={audioModalVisible}
+        onRequestClose={() => {
+          setAudioModalVisible(!audioModalVisible);
+        }}
+      >
+        <View style={styles.modal}>
+          <View style={{ alignItems: 'center', marginTop: 305 }}>
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: remove inline styling

File: src/screens/PlayScreen/Play.tsx [hunk 10]
```diff
@@ -1,69 +1,390 @@
...
+              >
+                <Text style={styles.inset_text2}>OK</Text>
+              </Pressable>
+            </View>
+          </View>
+        </View>
+      </Modal>
+
       <View
         style={{
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: remove inline styling

File: src/screens/PlayScreen/Play.tsx [hunk 11]
```diff
@@ -1,69 +1,390 @@
...
+          width: '100%',
+          marginTop: 20,
         }}
       >
-        <Icon type="dropdown" />
-        <Text style={styles.header_text}>
-          Green Colonization: An Interview with Maja Kristine Jama
-        </Text>
+        <View
+          style={{
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: remove inline styling

File: src/screens/PlayScreen/Play.tsx [hunk 12]
```diff
@@ -1,69 +1,390 @@
...
+            justifyContent: 'flex-start',
+            width: 336,
+          }}
+        >
+          <Icon type="dropdown" />
+          <Text style={styles.header_text}>{hardcodedResponse.title}</Text>
+        </View>
       </View>
-      <View style={styles.container2}>
+      <View
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: remove inline styling

File: src/screens/PlayScreen/Play.tsx [hunk 13]
```diff
@@ -1,69 +1,390 @@
...
+        style={{
+          flexDirection: 'column',
+          alignItems: 'center',
+          width: '100%',
+          marginTop: 26,
+        }}
+      >
         <Image
           style={{
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: remove inline styling

File: src/screens/PlayScreen/Play.tsx [hunk 14]
```diff
@@ -1,69 +1,390 @@
...
           }}
+          source={{ uri: hardcodedResponse.thumbnail }}
         />
       </View>
-      <Text style={styles.title_text}>
-        Green Colonization: An Interview with Maja Kristine Jama
-      </Text>
-      <Text style={styles.author_text}>Shaldon Ferris</Text>
+      <View
+        style={{
```
Comment (david-qing) [inline src/screens/PlayScreen/Play.tsx]: inline styling

File: assets/icons.tsx (hunk 6 above)
Comment (david-qing) [inline assets/icons.tsx]: fix colors

File: src/screens/PlayScreen/styles.tsx [hunk 15]
```diff
@@ -15,36 +15,117 @@ export default StyleSheet.create({
...
+  footer_container2: {
+    flexDirection: 'row',
+    justifyContent: 'flex-end',
+    width: 336,
+    marginBottom: 30,
+  },
+  image: {
+    height: 336,
+    width: 336,
+    backgroundColor: '#D9D9D9',
```
Comment (david-qing) [inline src/screens/PlayScreen/styles.tsx]: make this color from our colors styling; should be in styles/colors.ts

File: src/screens/PlayScreen/styles.tsx
Comment (david-qing) [inline src/screens/PlayScreen/styles.tsx:1]: for all values here, try to keep the values to multiples of ideally 8, but 4 if needed. we can fix this up later too, its a small change

File: src/screens/PlayScreen/styles.tsx [hunk 16]
```diff
@@ -15,36 +15,117 @@ export default StyleSheet.create({
...
+  },
+  modal: {
+    backgroundColor: 'rgba(217, 217, 217, 0.6)',
+    height: '100%',
+    width: '100%',
+  },
+  inset: {
+    backgroundColor: '#FFF',
```
Comment (david-qing) [inline src/screens/PlayScreen/styles.tsx]: use color value

File: src/screens/PlayScreen/styles.tsx [hunk 17]
```diff
@@ -15,36 +15,117 @@ export default StyleSheet.create({
...
+    width: 336,
+    marginBottom: 30,
+  },
+  image: {
+    height: 336,
+    width: 336,
+    backgroundColor: '#D9D9D9',
+  },
+  modal: {
+    backgroundColor: 'rgba(217, 217, 217, 0.6)',
```
Comment (david-qing) [inline src/screens/PlayScreen/styles.tsx]: use color value if possible, this one might not be possible

File: src/screens/PlayScreen/styles.tsx [hunk 18]
```diff
@@ -15,36 +15,117 @@ export default StyleSheet.create({
...
+  header_text1: {
     fontSize: 13,
-    width: '90%',
-    paddingLeft: '15%',
+    width: 291,
+    marginLeft: 20,
     textAlign: 'left',
-    color: Colors.textPrimary,
+    fontWeight: '500',
+    color: '#000000',
```
Comment (david-qing) [inline src/screens/PlayScreen/styles.tsx]: use color here

File: src/screens/PlayScreen/styles.tsx [hunk 19]
```diff
@@ -15,36 +15,117 @@ export default StyleSheet.create({
...
+  },
+  header_text2: {
+    fontSize: 17,
+    width: 291,
+    marginTop: -10,
+    marginLeft: 20,
+    textAlign: 'left',
+    fontWeight: '400',
+    color: '#000000',
```
Comment (david-qing) [inline src/screens/PlayScreen/styles.tsx]: color here

File: src/utils/AudioUtils.tsx [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+import { Audio, InterruptionModeAndroid, InterruptionModeIOS } from 'expo-av';
+import { AudioPlayerState } from '../AudioContext';
+
+export async function toggleAudio(
+  audio: AudioPlayerState,
+  setAudio: Dispatch<SetStateAction<AudioPlayerState>>,
+) {
+  const result = await audio.soundRef.getStatusAsync();
+  if (result.isLoaded) {
+    if (result.isPlaying === false) {
```
Comment (david-qing) [inline src/utils/AudioUtils.tsx]: could be useful to set in the other direction (if isplaying do some logic. else, do other logic)

**Review by david-qing — CHANGES_REQUESTED** [review]
code cleanliness needs work here

File: src/utils/AudioUtils.tsx (hunk 20 above)
Comment (kbrgl) [reply src/utils/AudioUtils.tsx]: Also, you don't have to do `result.isPlaying === false`, just do`!result.isPlaying`

Comment (kbrgl) [reply src/utils/AudioUtils.tsx]: But like david said it should be if (result.isPlaying), else .... Makes the logic cleaner

File: src/screens/PlayScreen/styles.tsx (hunk 17 above)
Comment (david-qing) [reply src/screens/PlayScreen/styles.tsx]: add this line into the styles/colors.ts file:

translucentBackground: '#d9d9d999',

File: src/utils/AudioUtils.tsx [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+        isPlaying: true,
+      }));
+      await audio.soundRef.playAsync();
+    }
+  }
+}
+
+export async function rewindAudio(
+  audio: AudioPlayerState,
+  // setAudio: Dispatch<SetStateAction<AudioPlayerState>>,
```
Comment (kbrgl) [inline src/utils/AudioUtils.tsx]: Get rid of unnecessary comment

**Comment by kbrgl** [comment]
Great work on this PR :))

**Review by david-qing — APPROVED** [review]
looks good to merge!

File: src/screens/PlayScreen/Play.tsx (hunk 5 above)
Comment (akshaynthakur) [reply src/screens/PlayScreen/Play.tsx]: yeah this causes bugs just because of the way the sound object is loaded

**Review by david-qing — APPROVED** [review]
Looks good to merge :)
